### ISSUE TITLE
Make imports in Python consistent with the style guide

### ISSUE THREAD

**Issue by gimite** [google/personfinder#615]
Our current code base has many imports in Python not following this style guide:
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#22-imports

* We shouldn't import wildcard like this: https://github.com/google/personfinder/blob/master/app/view.py#L18
* We shouldn't import individual functions.

These cases should be fixed to import packages/modules instead.

**Comment by vintol** [comment]
I will take a shot at this.

**Comment by gimite** [comment]
> > I would like to take a shot at this.

That would be great, thanks!

Let me take examples to better understand your proposal.

https://github.com/google/personfinder/blob/master/app/view.py contains:
```python
from model import *
...
person = Person.get(self.repo, self.params.id)
```
and https://github.com/google/personfinder/blob/master/app/create.py contains:
```python
from photo import create_photo, PhotoError
...
note_photo, note_photo_url = \
    create_photo(self.params.note_photo, self)
```

How does the code look like by applying your proposed change? What I imagined was:

```python
import model
...
person = model.Person.get(self.repo, self.params.id)
```

```python
import photo
...
note_photo, note_photo_url = \
    photo.create_photo(self.params.note_photo, self)
```

> Whoa, grepped the import lines.
> That's some 4000 imports in whole code.
> Might take some time.
> Especially where I have to find the modules actually being used
> where `from x import *` is used.
> 
> **Let me know if there is a time limit or deadline or something for this.**
> I am doing this in my spare time.

There is no time limit or deadline, so take your time. And you don't need to do everything at once (and you probably shouldn't). It would probably be best to first send a pull request to only fix a single file (e.g., app/create.py) to check if the change matches our intention, and apply it to other files later.

Just FYI I found a bit tricky example:
```python
photo, photo_url = create_photo(self.params.photo, self)
```
In this case the variable must be also renamed e.g.:
```python
person_photo, person_photo_url = photo.create_photo(self.params.photo, self)
```
To avoid collision with the module name. Something worth keeping in mind.

**Comment by vintol** [comment]
What I meant was this : 
```python
from model import *
...
person = Person.get(self.repo, self.params.id)
```

This could be done in one of 2 ways:

1.  The same thing @gimite did in post above.

```python
import model
...
person = model.Person.get(self.repo, self.params.id))
```
> This is obviously the one to aim for but it might require more attention during this change.

- Any new commit mades during this process of updating import statement in entire codebase would have to be made with this in mind.
- Names of some functions and variables might have to be changed that conflict with the name of the package.
- This change would have to reflect in every file that uses or imports that function.
- Morevoer, also in the new code that is being pushed into the codebase.

Take a look at this file : 



2.  Other way
```python
from model import Person,[Some Other Module],[Some Another Module],[Etc ...]
...
person = Person.get(self.repo, self.params.id))

``` 
> The advantage of the 2nd way is that is does not change anything else than the import code.

Tommorrow, 
I intend to show case a file with reformatted code : app/view.py

A quick question though,

in view.py I see
Line 18 `from photo import create_photo, PhotoError`
but that is not used anywhere in the file and is not required.
Am I mistaken ?

**Comment by gimite** [comment]
Thanks for the clarification!

> ```python
> from model import Person,[Some Other Module],[Some Another Module],[Etc ...]
> ...
> person = Person.get(self.repo, self.params.id))
> ```

This still violates our style guide:
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#22-imports
> Use import statements for packages and modules only, not for individual classes or functions.

So I would prefer to go with the first option even though it is more work.

> A quick question though,
> 
> in view.py I see
> Line 18 `from photo import create_photo, PhotoError`
> but that is not used anywhere in the file and is not required.
> Am I mistaken ?

It is true for this particular case, and you would also see a lot of unused imports like this unfortunately :( Feel free to (and you are welcomed to) clean up unused imports as well when you find them.

**Comment by vintol** [comment]
The [file](https://gist.github.com/vintol/302d303af78150c4d1ce1910dc04333e) that I promised.

> I am having some problems with the developement environment.

At first I thought,
for the imports like `from model import *`

I would have to search every class and function there is in model.py in the present file.
But then I was hoping for another way, something like

```python

Default = dir()
from model import *
New = dir()
for i in New:
    if i not in Default:print(i)

```
This is premitive, i know. But it's just to give you an idea.
Then maybe look for that imported functions in the present file manually via search or maybe write some more code.

This would save much time.

But, I get this error, when importing the python file
Using the appengine directly downloaded in the PATH
ModuleNotFoundError: No module named 'google'
Using the SDK installment of appengine in the PATH:
ModuleNotFoundError: No module named 'appimage'
But I was able to run the entire program via
`env APPENGINE_DIR="/home/***/Downloads/google_appengine/" tools/gae run app --port=8000`

Anyhow,
After some while I gave that up.
And started writing a script that relies on text parsing.
Based on the keywords like def, class and indentation.
I get a list of func and class, (variables not imeplemented yet),
then look for those names in the file that I am currently working on.

It still needs work but it will greatly reduce the effort of going through the entire files.

Few things on which I need some feedback : 

>  `from django.utils.translation import ugettext as _`
What to do here ? 

Rename that ?

1 ]  `import django.utils.translation.gettext as _`

2 ] `import django.utils.translation.gettext`

3] `import django.utils.translation`

4] One can not go any higher right? `import django.utils` won't work. 


> Also this one, `from google.appengine.api import datastore_errors`

**Comment by nworden** [comment]
> But, I get this error, when importing the python file
> Using the appengine directly downloaded in the PATH
> ModuleNotFoundError: No module named 'google'

To be clear, you mean when including APPENGINE_DIR on the path, same as when you run the app? Could you let me know what version of the SDK you're using (it should be the first line of output if you run `cat $APPENGINE_DIR/VERSION`)? And where you're running it from and where your script is located relative to the app directory?

> from django.utils.translation import ugettext as _
> What to do here ?

IMHO this is the sort of situation where maybe it's not worth following the style guide (_('user-facing') string seems more readable than the alternatives), but I'll yield to Hiroshi. If we do want to stick to the style guide, one option to minimize the length of the calls would be to import the translation module like this:
`import django.utils.translation as t`
so that the calls could just be `t.ugettext`.

**Comment by gimite** [comment]
> IMHO this is the sort of situation where maybe it's not worth following the style guide

I agree that we should consider this as an exception and leave it as is, because I believe it's a common pattern for this particular module. Sorry I forgot to mention that earlier.

**Comment by vintol** [comment]
I updated my earlier comment. Maybe take a look at that for other things.
@nworden :

# Devlopment Environment

For convenience, I am showing just the directory name here,
In reality, I am using full path wherever applicable.

## Python2 + pips

Using python2 env setup using miniconda
Installed all the requirements.txt plus others mentioned in the guide to set up dev env.

conda set up is in a directory `tes`. I have created a symlink of that python2 executable to my /usr/local/bin called py2.
Running everything using that.

## G SDK

Google SDK v240.0.0
set up in it's own dir outside the tree of `tes` called `gsdk`.
I changed the script of `gsdk/bin/gcloud` to always use `python=py2`

Installed the app engine python and extras.
> I remember sdk modified .bashrc but whole time I am using fish shell.

Trying something new here, nut here's what I did before. 
prepended the PATH with `gsdk`.

cd to the personfinder.
```python
> py2
import google # works fine but Nothing in dir(google).
import google.appimage #ERROR : No module named `appimage` 
```
## G App Engine

Directly downloaded and extrected.
to a dir `appengine`

prepended the PATH with `appengine`.
cd to the personfinder.
```python
> py2
import google #ERROR : No module named `google` 
import google.appimage  #ERROR : No module named `google` 
```

## Run personfinder

prepended the PATH with `gsdk`.
cd to the personfinder.
```
Have not tried that.
```

prepended the PATH with `appengine`.
cd to the personfinder.
```fish
env APPENGINE_DIR="appengine" tools/gae run app --port=8000
```
That works fine.

**Comment by vintol** [comment]
I have an Important question. 

> How do we integrate the changes being made ?

Do I just commit to the main branch, can I even do that ?
OR keep working on the forked branch or something else ... ?

**Comment by gimite** [comment]
You can just send pull requests. We can review and merge them into the master branch.

**Comment by vintol** [comment]
Okay. Doing that.
I have started working.

I am Removing the first comment I made and using it as a sort of TODO board and a Progress bar for this Issue.

**Comment by vintol** [comment]
In Model.py

`from google.appengine.ext import db`
135 refrences to db.

`from google.appengine import ext` or leave as it is ?

**Comment by gimite** [comment]
https://cloud.google.com/appengine/docs/standard/python/refdocs/google.appengine.ext.db?hl=en
`db` is a package, so you can leave it as is.

It is sometimes unclear if the symbol is a package/module or a variable. But I believe you can consult documentation or read code of the library to distinguish them.
